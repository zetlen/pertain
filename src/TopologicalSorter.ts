import PertainError from './PertainError';
/* tslint:disable */

/**
 * Topologically sort a list of items of any type, given a list of items and
 * a map of outward relations between those items. (Could be a function, but
 * it's a class for testability.)
 *
 * Credit to https://github.com/marcelklehr/toposort/ for a great
 * implementation I had to adapt to be faster for our particular use case.
 *
 * That code appears here courtesy of the MIT license.
 */
export default class TopologicalSorter<T> {
  private outgoingFrom: (node: T) => T[];
  constructor(outgoingFrom: (node: T) => T[]) {
    this.outgoingFrom = outgoingFrom;
  }
  public sort(nodes: T[]): T[] {
    let cursor = nodes.length;
    const sorted: T[] = new Array(cursor);
    const visited: { [key: number]: boolean } = {};
    let i = cursor;
    // Better data structures make algorithm much faster.
    const nodesHash = this.makeNodesHash(nodes);

    const { outgoingFrom } = this;

    while (i--) {
      if (!visited[i]) visit(nodes[i], i, new Set());
    }

    return sorted;

    function visit(node: T, i: number, predecessors: Set<T>): void {
      if (predecessors.has(node)) {
        throw new PertainError(`Cyclic dependency on ${node}`);
      }

      if (visited[i]) return;
      visited[i] = true;

      const outgoing = outgoingFrom(node);

      if ((i = outgoing.length)) {
        predecessors.add(node);
        do {
          let child = outgoing[--i];
          visit(child, nodesHash.get(child) as number, predecessors);
        } while (i);
        predecessors.delete(node);
      }

      sorted[--cursor] = node;
    }
  }
  private makeNodesHash(arr: T[]) {
    const res = new Map<T, number>();
    for (let i = 0, len = arr.length; i < len; i++) {
      res.set(arr[i], i);
    }
    return res;
  }
}
