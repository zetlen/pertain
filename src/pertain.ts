import makeDebug from 'debug';
import ExplicitDependencySet from './ExplicitDependencySet';
import PackageJson from './PackageJson';
import resolver from './resolverFactory';

const debug = makeDebug('pertain:main');

export type DependencyFinder = (
  found: string[],
  packageJson: PackageJson,
  rootDir: string,
  subject: string
) => string[];

const depsAndDevDeps = (pkg: PackageJson): string[] => {
  // Merging the two dependency sets that we look at will dedupe them.
  // We don't care whether it comes from devDependencies or dependencies.
  // Both are relevant, because many applications with a build step compile
  // code from devDependencies.
  const allDependencyNames = Object.keys({
    ...pkg.dependencies,
    ...pkg.devDependencies,
  });
  allDependencyNames.push('./'); // rootDir too
  return allDependencyNames;
};

/**
 * A Node package which declares in its `package.json` file that it contains
 * a script module which pertains to a particular subject. Use this object
 * to require the pertaining module and/or report the name of the subscriber.
 */
export interface Pertaining {
  /**
   * The name of the package..
   * @example "@org/package-name"
   */
  name: string;
  /**
   * The real path of the package root.
   * @example "/Users/you/project/node_modules/@org/package-name"
   */
  modulePath: string;
  /**
   * The filesystem path of the package's root directory.
   * @example "/var/www/venia/node_modules/@org/package-name"
   */
  path: string;
  /**
   * The subject that was originally passed to `pertain()` to produce this
   * object.
   * @example "pwa.webpack"
   */
  subject: string;
}

/**
 * Caches the ExplicitDependencySet calculated for a given project root.
 * It should not change during the lifetime of this process, so there's no
 * point in recalculating it.
 */
const dependencySetCache = new Map<string, ExplicitDependencySet>();

/**
 * Query the direct dependencies of the Node project at `rootDir` for all
 * packages which have a particular `package.json` property. Return them in
 * peerDependency order.
 */
function pertain(
  rootDir: string,
  subject: string,
  getDependencies: DependencyFinder = (found) => found
): Pertaining[] {
  let depSet = dependencySetCache.get(rootDir);
  if (!depSet) {
    debug('no cached depset for %s', rootDir);
    // A convenience function which can be replaced with an alternate resolver
    // algorithm.
    const resolve = resolver(rootDir);
    const packageJson = new PackageJson(rootDir);
    const allDependencyNames = getDependencies(
      depsAndDevDeps(packageJson),
      packageJson,
      rootDir,
      subject
    );
    debug('%s allDependencyNames %s', rootDir, allDependencyNames);
    depSet = new ExplicitDependencySet(resolve, allDependencyNames);
    dependencySetCache.set(rootDir, depSet);
  }
  return depSet.pertaining(subject).map((dep) => ({
    name: dep.name,
    modulePath: dep.modulePath,
    path: dep.pertains(subject) as string,
    subject,
  }));
}

/**
 * Clear out the cache of dependencies that have already been detected and
 * loaded. Use if the dependency graph changes and you want to "hot reload"
 * functionality. Or, for testing.
 */
pertain.clearCache = () => dependencySetCache.clear();

export default pertain;
