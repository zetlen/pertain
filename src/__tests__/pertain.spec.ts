import path from 'path';
import pertain from '../';
import TopologicalSorter from '../TopologicalSorter';

const projectPath = path.join(__dirname, '__fixtures__/cool-cactus');

test('pertains to "germane.identify" in dep order to list ingredients in a cool cactus', () => {
  const identifiers = pertain(projectPath, 'germane.identify');
  const ingredients = identifiers
    .map(i => `- ${i.name} adds ${require(i.path)()}`)
    .join('\n');
  expect(ingredients).toMatchSnapshot();
});

test('runs the same sequence without invoking sort twice', () => {
  pertain.clearCache();
  const sort = jest.spyOn(TopologicalSorter.prototype, 'sort');
  pertain(projectPath, 'germane.identify');
  pertain(projectPath, 'germane.identify');
  expect(sort).toHaveBeenCalledTimes(1);
  sort.mockRestore();
});

test('pertains to "germane.draw" in dep order to draw the cool cactus', () => {
  const picture = pertain(projectPath, 'germane.draw').reduce(
    (drawing, artist) => require(artist.path)(drawing),
    ''
  );
  expect(picture).toMatchSnapshot();
});

test('throws informative error if a dep declares an invalid pertainer', () => {
  expect(() => pertain(projectPath, 'lies')).toThrow('could not find');
});

test('throws informative error if circular deps are detected', () => {
  expect(() => pertain(projectPath, 'humpty')).toThrow('Cyclic');
});

test('accepts custom DependencyFinder', () => {
  pertain.clearCache();
  const allTalk = pertain(projectPath, 'germane.identify', () => [
    'cactus-attitude'
  ])
    .map(dep => require(dep.path)())
    .join('');
  expect(allTalk).toBe('a bad attitude');
});
