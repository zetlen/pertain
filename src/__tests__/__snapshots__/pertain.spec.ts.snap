// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pertains to "germane.draw" in dep order to draw the cool cactus 1`] = `
"
@============================================@
|                                            |
|                   / '-' }                  |
|                  ;       ;                 |
|               /'-|       |-'}              |
|              |   |_______K   |             |
|              }   '-------'   /             |
|               '.___.....___.'              |
|                  |     ; |                 |
|                 _|;__;__.|_                |
|                |     Y     |    .--.       |
|       .--.      X__.'^'.__/                |
|                  |    ;  |      .--.       |
|       .--.       | ;  .  |     /;   }      |
|      /   ;}      |_  ;  _|     |  ' |      |
|      | ;  |      {  ;  * |     |;   |      |
|      |'   |      {    .  }     | ;  |      |
|      |  ; |      {  *    }     |    |      |
|      |;   |      ;'-.__.'|     |:  ;|      |
|      | ;  }      |;  ;   |_____/ ;  |      |
|      |   '.'-----'      ' -_   .'   /      |
|      }  '.   - _  ' ;  ;  _  -    .'       |
|       '.   -     - ;  ;   .------'         |
|         '--------.      ;|                 |
|                  |;  ,   |                 |
|                  |     ; |                 |
|                  |. ;    |                 |
|                  | :    :|                 |
|                  |;   ;  |                 |
|                  |     ; |                 |
|                  |     ; |                 |
|                  |. ;    |                 |
|                  | :    :|                 |
|                  |   .   |                 |
|                  |;   ;  |                 |
|                                            |
|          ceci n'est pas un cactus          |
|                                            |
@============================================@"
`;

exports[`pertains to "germane.identify" in dep order to list ingredients in a cool cactus 1`] = `
"- @garden/cactus adds a cactus in the garden
- cactus-attitude adds a bad attitude
- cactus-hat adds a hat
- cactus-glasses adds a pair of sunglasses"
`;

exports[`provides info about pertaining dependencies, in dependency order 1`] = `
[
  {
    "modulePath": "node_modules/@garden/cactus",
    "name": "@garden/cactus",
    "path": "node_modules/@garden/cactus/draw.js",
    "subject": "germane.draw",
  },
  {
    "modulePath": "node_modules/cactus-hat",
    "name": "cactus-hat",
    "path": "node_modules/cactus-hat/hatify.js",
    "subject": "germane.draw",
  },
  {
    "modulePath": "node_modules/cactus-glasses",
    "name": "cactus-glasses",
    "path": "node_modules/cactus-glasses/insert-glasses.js",
    "subject": "germane.draw",
  },
  {
    "modulePath": "node_modules/treachery",
    "name": "treachery",
    "path": "node_modules/treachery/index.js",
    "subject": "germane.draw",
  },
]
`;
